### PR TITLE
Bug 1469561 - Reshape Settings

### PR DESCRIPTION
**Pull request by ghickman**
This organises our core settings into three groups: Django, Third Party, & Treeherder (project level).  Internally to the latter two the settings have been grouped into either App or related sections in alphabetical order (to aid scanning the file).  Any settings only used in one place have also been moved to the respective usage site.

### CODE REVIEW

File: treeherder/config/settings.py [hunk 1]
```diff
@@ -327,76 +399,73 @@
...
+# Debug Toolbar
+if ENABLE_DEBUG_TOOLBAR:
+    # django-debug-toolbar requires that not only DEBUG be set, but that the request IP
+    # be in Django's INTERNAL_IPS setting. When using Vagrant, requests don't come from localhost:
+    # http://blog.joshcrompton.com/2014/01/how-to-make-django-debug-toolbar-display-when-using-vagrant/
+    # If the Vagrant IPs vary by platform or if there isn't a consistent IP when we switch to Docker,
+    # we'll have to do: https://github.com/jazzband/django-debug-toolbar/pull/805#issuecomment-240976813
+    INTERNAL_IPS = ['127.0.0.1', '10.0.2.2']
+
+if ENABLE_DEBUG_TOOLBAR:
```
Comment (edmorley) [inline treeherder/config/settings.py]: This conditional is duplicated above. Also, in the Django section immediately after `INSTALLED_APPS` there is an `if DEBUG` that adds the same app. I think the reason we used the additional environment variable before, was that django debug toolbar is in `requirements/dev.txt`, so if `DEBUG` was enabled on eg prototype for some reason, then the dependency wouldn't be found, and so the app not start. However I really doubt using DEBUG is something we ever would need to (or should) do, even on the prototype instance -- so let's just ditch `ENABLE_DEBUG_TOOLBAR` (incl from `env.sh`).

File: treeherder/config/settings.py [hunk 2]
```diff
@@ -327,76 +399,73 @@
...
+# Rest Framework
 REST_FRAMEWORK = {
-    'DEFAULT_PARSER_CLASSES': (
-        'rest_framework.parsers.JSONParser',
-    ),
-    'DEFAULT_FILTER_BACKENDS': (
-        'rest_framework_filters.backends.DjangoFilterBackend',
-    ),
-    'DEFAULT_RENDERER_CLASSES': (
+    'ALLOWED_VERSIONS': ['1.0'],
```
Comment (edmorley) [inline treeherder/config/settings.py]: This commit changes the d-r-f settings from tuples to lists. Whilst both are valid, the reason we'd used tuples is since that's what the d-r-f docs and defaults use:
https://github.com/encode/django-rest-framework/blob/3.8.2/rest_framework/settings.py#L31

File: treeherder/config/settings.py [hunk 3]
```diff
@@ -84,14 +82,72 @@
...
+        'LOCATION': REDIS_URL,
+        'OPTIONS': {
+            # Override the default of no timeout, to avoid connection hangs.
+            'SOCKET_CONNECT_TIMEOUT': 5,
+        },
+    },
+}
+
+# Internationalization
+LANGUAGE_CODE = 'en-us'
```
Comment (edmorley) [inline treeherder/config/settings.py]: `LANGUAGE_CODE` wasn't set before - it is a new addition in this commit. The Django default value is `en-us` ([docs](https://docs.djangoproject.com/en/1.11/ref/settings/#language-code)) - we've generally avoided repeating any settings in this file that are using the defaults - though open to thoughts about this :-)

File: treeherder/config/settings.py [hunk 4]
```diff
@@ -84,14 +82,72 @@
...
+TIME_ZONE = "UTC"
+USE_I18N = False
+USE_L10N = True
+USE_TZ = False
```
Comment (edmorley) [inline treeherder/config/settings.py]: `USE_TZ` also wasn't set before, and defaults to `False` (so the current setting is repeating the default) -- however I seem to remember we should really be using `True`, and as such we might want this addition as a reminder? If not, then let's remove it too.

File: treeherder/model/management/commands/cycle_data.py [hunk 5]
```diff
@@ -10,6 +10,16 @@
...
 
+env = environ.Env()
+
+# Default to retaining data for ~4 months.
+DATA_CYCLE_DAYS = env.int("DATA_CYCLE_DAYS", default=120)
+
+# Determines the number of jobs we try to delete per iteration when
+# cycling data
+DATA_CYCLE_CHUNK_SIZE = env.int("DATA_CYCLE_CHUNK_SIZE", default=100)
+DATA_CYCLE_SLEEP_TIME = env.int("DATA_CYCLE_SLEEP_TIME", default=0)
```
Comment (edmorley) [inline treeherder/model/management/commands/cycle_data.py]: We haven't used any of these `DATA_CYCLE_*` settings for a while (previously different expiration times were set per environment), so let's just remove the options and transpose the defaults (and comments) below.

File: treeherder/config/settings.py [hunk 6]
```diff
@@ -327,64 +399,71 @@
...
-ALLTHETHINGS_URL = "https://secure.pub.build.mozilla.org/builddata/reports/allthethings.json"
-TASKCLUSTER_INDEX_URL = 'https://index.taskcluster.net/v1/task/gecko.v2.%s.latest.taskgraph.decision'
-TASKCLUSTER_RUNNABLE_JOBS_URL = 'https://public-artifacts.taskcluster.net/{task_id}/0/public/runnable-jobs.json.gz'
+# TREEHERDER
 
-# the amount of time we cache bug suggestion lookups (to speed up loading the bug
-# suggestions or autoclassify panels for recently finished jobs)
-BUG_SUGGESTION_CACHE_TIMEOUT = 86400
+# Autoclassification
+AUTOCLASSIFY_JOBS = env.bool("AUTOCLASSIFY_JOBS", default=True)
```
Comment (edmorley) [inline treeherder/config/settings.py]: We can probably remove this

File: treeherder/config/settings.py [hunk 7]
```diff
@@ -393,70 +472,36 @@
 BUILDAPI_RUNNING_CHUNK_SIZE = 500
 BUILDAPI_BUILDS4H_CHUNK_SIZE = 500
 
+BUILDAPI_PENDING_URL = "https://secure.pub.build.mozilla.org/builddata/buildjson/builds-pending.js"
+BUILDAPI_RUNNING_URL = "https://secure.pub.build.mozilla.org/builddata/buildjson/builds-running.js"
+BUILDAPI_BUILDS4H_URL = "https://secure.pub.build.mozilla.org/builddata/buildjson/builds-4hr.js.gz"
+ALLTHETHINGS_URL = "https://secure.pub.build.mozilla.org/builddata/reports/allthethings.json"
```
Comment (edmorley) [inline treeherder/config/settings.py]: Happy for most of these to be moved to where they are used, too (can be in another PR if preferred)

File: treeherder/config/settings.py (hunk 3 above)
Comment (ghickman) [reply treeherder/config/settings.py]: Django sets this in their [generated settings template](https://github.com/django/django/blob/master/django/conf/project_template/project_name/settings.py-tpl#L106) and IIRC has done for many versions.

That being said!  I think you're right here and I'll remove them :)

File: treeherder/config/settings.py [hunk 8]
```diff
@@ -196,6 +218,58 @@
...
+# THIRD PARTY APPS
+
+# Auth0 setup
+AUTH0_DOMAIN = env('AUTH0_DOMAIN', default="auth.mozilla.auth0.com")
+AUTH0_CLIENTID = env('AUTH0_CLIENTID', default="q8fZZFfGEmSB2c5uSI8hOkKdDGXnlo5z")
+
+# Set the `X-Frame-Options` header, which forbids embedding of site pages in frames other than origin.
+# AUTH0 renewal opens the auth handler page in an invisible frame
+# hence requiring the need to support frames with same origin
+X_FRAME_OPTIONS = 'SAMEORIGIN'
```
[the remark this answers is not in the document]
Comment (ghickman) [reply treeherder/config/settings.py]: I think I mistakenly left this here because of the context of the comment.  However given it's currently set to the default value, should we remove it?

**Comment by ghickman** [comment]
Rebased changes into appropriate commits or added new ones for things like the full removal of settings.

Just the one query!

File: treeherder/model/management/commands/cycle_data.py [hunk 9]
```diff
@@ -10,6 +10,8 @@
                                      Repository)
 from treeherder.perf.models import PerformanceDatum
 
+env = environ.Env()
```
Comment (edmorley) [inline treeherder/model/management/commands/cycle_data.py]: django-environ is unused in this file now

File: treeherder/config/settings.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (edmorley) [reply treeherder/config/settings.py]: Yeah lets remove it (it exists in this file since it was previoulsy set to `DENY` but had to be adjusted when we switched to Auth0) - but I'd perhaps move some of the additional comment detail into the existing `DISALLOWED_USER_AGENTS` comment.

[the remark this answers is not in the document]
Comment (ghickman) [reply treeherder/config/settings.py]: Done in e714d7056

File: treeherder/model/management/commands/cycle_data.py (hunk 9 above)
Comment (ghickman) [reply treeherder/model/management/commands/cycle_data.py]: Fixed and rebased

File: treeherder/config/settings.py [hunk 10]
```diff
@@ -196,6 +215,56 @@
...
+    # made using Angular's `httpProvider` require access to the cookie.
+    'security.W017',
+    # We can't set X_FRAME_OPTIONS to DENY since renewal of auth token requires
+    # opening an invisible iframe with the same origin.
+    'security.W019'
+]
+
+# User Agents
+# User agents which will be blocked from making requests to the site.
+# Note: we use the default setting for X_FRAME_OPTIONS ('SAMEORIGIN') so Auth0
```
Comment (edmorley) [inline treeherder/config/settings.py]: This comment need to be slightly higher up, above the `security.W019` line and combined with the comment there.

Comment (ghickman) [reply treeherder/config/settings.py]: Ah, sorry I misread your last comment!

Comment (edmorley) [reply treeherder/config/settings.py]: Oh I actually wrote the wrong thing haha

Comment (ghickman) [reply treeherder/config/settings.py]: Rebased into the latest commit

Comment (edmorley) [reply treeherder/config/settings.py]: 👍
